cmds/legacy_boot/legacy_boot.go

```diff
@@ -0,0 +1,383 @@
+// Copyright 2012-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// legacy_boot allows to handover a system running linuxboot/u-root
+// to a legacy preinstalled operating system by replacing the traditional
+// bootloader path
+
+//
+// Synopsis:
+//	legacy_boot
+//
+// Description:
+//	If returns to u-root shell, the code didn't found a local bootable option
+//
+// Notes:
+//	The code is looking for boot/grub/grub.cfg file as to identify the
+//	boot option.
+//	The first bootable device found in the block device tree is the one used
+//	Windows is not supported (that is a work in progress)
+//
+// Example:
+//	legacy_boot -v 	- Start the script in verbose mode for debugging purpose
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/u-root/u-root/pkg/kexec"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+	"flag"
+	"syscall"
+)
+
+var verbose bool
+type options struct {
+        verbose bool
+}
+
+func check(e error) {
+	if e != nil {
+		panic(e)
+	}
+}
+
+func blk_devices_list(blkpath string, devpath string) []string {
+	var blk_devices []string
+	files, err := ioutil.ReadDir(blkpath)
+	if err != nil {
+		log.Fatal(err)
+	}
+	for _, file := range files {
+		device_entry, err := ioutil.ReadDir(blkpath + file.Name() + devpath)
+		if err == nil {
+			blk_devices = append(blk_devices, device_entry[0].Name())
+		}
+	}
+	return (blk_devices)
+}
+
+// Current support is limited to Hard disk devices and USB devices
+
+func check_for_bootable_mbr(path string) int {
+	var b511, b510 byte
+	f, err := os.Open(path)
+	check(err)
+	b1 := make([]byte, 512)
+	f.Read(b1)
+	b511 = b1[511]
+	b510 = b1[510]
+	f.Close()
+	if ((b511 == 0xaa) && (b510 == 0x55)) == true {
+		return 1
+	}
+	return 0
+}
+
+// Return all devices attached to a specific name like /dev/sdaX where X can move from 0 to 127
+// FIXME no support for devices which are included into subdirectory within /dev
+
+func get_device_part_list(path string) []string {
+	var return_value []string
+	files, err := ioutil.ReadDir("/dev/")
+	if err != nil {
+		log.Fatal(err)
+	}
+	for _, file := range files {
+		if strings.HasPrefix(file.Name(), path) {
+			// We shall not return full device name
+			if file.Name() != path {
+				// We shall check that the remaining part is a number
+				return_value = append(return_value, file.Name())
+			}
+		}
+	}
+	return return_value
+}
+
+// Return all block file system supported by the linuxboot kernel
+
+func get_supported_filesystem() []string {
+	var return_value []string
+	file, err := os.Open("/proc/filesystems")
+	check(err)
+	defer file.Close()
+	scanner := bufio.NewScanner(file)
+	if err := scanner.Err(); err != nil {
```

Comment:
check for scanner.Err() after the scanner.Scan() loop ends. Err() is only set if during the scanning an error is encountered -- after Scan returns false, Err is nil if the file ended and Err is non-nil if Scan encountered some other error (I/O, etc).